### ISSUE TITLE
Baikal sync not working

### ISSUE THREAD

**Issue by ronschaeffer** [kdw2060/hassio-addons#12]
Thanks very much for the add on. I am trying to it working with self hosted Baikal without success. I can sync with the Baikal calendar using desktop applications like 1Calendar.

Add on log:

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] done.
[services.d] starting services
[services.d] done.
added 62 packages, and audited 63 packages in 7s
21 packages are looking for funding
  run `npm fund` for details
found 0 vulnerabilities
error: Request failed with status code 500[object Object]
Calendar(s) queried at: Tue Apr 12 2022 11:00:00 GMT+0100 (British Summer Time)
postEventsAllCalendars error: TypeError: Cannot read property 'length' of undefined
CRON-Task: Events posted to sensor(s) at: Tue Apr 12 2022 11:00:00 GMT+0100 (British Summer Time)
```

Nothing at all appears in the Baikal log. 

I noticed that with version 0.200 support for Baikal calldav calendars is added. Presumably that means that the following from the docs no longer applies:  "The add-on uses the jcal format querying the server. Make sure this plugin is installed in your Baikal instance. See [Baikal Plugin Info](https://sabre.io/dav/ics-export-plugin/) for more information."

Add on configuration:
```
locale: en-gb
fetchDays: 365
fetchCRON: '*/30 * * * *'
calendarList:
  - calName: twickenhamevent
    calType: baikal
    username: xxxxx
    password: xxxxx
    caldavUrl: http://xxx.xxx.xxx.xxx:xxxx/dav.php/calendars/136worple/twickenhamevent
```
One thing I notice versus the  documentation is that I'm not using SSL whereas the documenation notes using an https address.

Any ideas? Thank you.

**Comment by ronschaeffer** [comment]
Server and plug-in info [here](https://1drv.ms/u/s!AtYGFztGDGYlkq1Thmm68SyqzunBiA?e=3FwPGJ)

**Comment by kdw2060** [comment]
Hi, no those docs were updated for version 0.200 and that information is new since then. Thus for Baikal that plugin is required, because the parsing of the data depends on it.

But the error log show a status code 500 error, which suggests another issue. That's an error you get when an endpoint (your caldav server) sends 'internal server error'. For starters i'd try an https url. 
If that doesn't work: I don't run Baikal myself and have no experience with it. User [@cellerich](https://github.com/cellerich) wrote that code and can probably help you better to debug the issue.

**Comment by cellerich** [comment]
Hi Ron

Just noticed that you have the `locale:` keyword in the wrong position. It belongs to the calendar:

```yaml
fetchDays: 365
fetchCRON: '*/30 * * * *'
calendarList:
  - calName: twickenhamevent
    calType: baikal
    locale: en-gb
    username: xxxxx
    password: xxxxx
    caldavUrl: http://xxx.xxx.xxx.xxx:xxxx/dav.php/calendars/136worple/twickenhamevent
```

And like @kdw2060 mentioned, the Baikal `jcal-query` plugin is needed!

**Comment by ronschaeffer** [comment]
Thanks, @cellerich . I'll try that.

Just FYI, the default configuration example on the add-on configuration page has the locale outside of the calendar section:

```
locale: nl-be
fetchDays: 365
fetchCRON: '*/30 * * * *'
calendarList:
  - calName: null
...
```
As does the example in DOCS.md:
```
fetchDays: 365
fetchCRON: '*/30 * * * *'
locale: nl-be
calendarList:
  - calName: someName
 ...
```

**Comment by cellerich** [comment]
@ronschaeffer  you are right, `locale` is fetched only once at the beginning of the code. Nothing to do with a specific calendar.
Just checked against my config and I have the `locale` option within the calendar object and it works, weird.

Anyway, the most likely problem is maybe the missing `jcal-query` plugin in your installation.

Have you tried to use `calType: caldav`? This should work with a Baikal server as well, just using a different API to fetch the data. Just make sure to add a `/` at the end of your `caldavUrl` with this type!!

**Comment by ronschaeffer** [comment]
Successwith `calType: caldav`. I did not need to set up access using SSL.

`calType: baikal` is still giving the same error `postEventsAllCalendars error: TypeError: Cannot read property 'length' of undefined`, even though I appear to have the` jcal-query` plug-in ([this one](https://sabre.io/dav/ics-export-plugin/l), right?) installed as shown when accessing the server URL through a browser:

![image](https://user-images.githubusercontent.com/12379923/163221853-ff6db018-0a5e-4d35-a1c6-3b357cf2abd4.png)

Since `calType: caldav` is working for me, I can close the issue if you like. Otherwise, I'd also be happy to keep testing with `calType: baikal` if it's helpful.

**Comment by kdw2060** [comment]
Judging from the code I don't think that's the right plugin. The plugin exports the calendar data as json in stead of an ics file, making it easier to work with the data and offloading possible data conversion issues to Baikal itself.

If you no longer get the Request failed ...500 (or other status code error) with the Baikal setting, but only the 'cannot read property length', that should mean any of these things:
- the async javascript tripped up > just wait till the next fetch round and then you should have your data
- your calendar contains no events for the given period
- there's something wrong with the received data object

As for the locale setting, yeah I moved that out of the per calendar settings as it seemed more logical to treat it as a shared setting.

**Comment by ronschaeffer** [comment]
Hmm. It's the same plug as is[ linked in the docs](https://sabre.io/dav/ics-export-plugin/). There is a `accept=jcal` option mentioned there that appears to work for me. in Postman. 

But, I can't get it to work using all of the options I see in the code here:

```
    // create date objects for Baikal fetch call
    let startTimestamp = Math.floor(new Date().getTime()/1000);
    let fetchDateFromNow = new Date();
    fetchDateFromNow.setDate(fetchDateFromNow.getDate() + fetchDays);
    let endTimestamp = Math.floor(fetchDateFromNow.getTime()/1000);

    // fetch Calendar objects for Baikal CALDAV
    if (options.calendarList[i].calType === "baikal") {
      let eventUris = [];
      const caldavUrl = options.calendarList[i].caldavUrl;
      const toEncode = username + ":" + pass;
      const buff = Buffer.from(toEncode, "utf-8");
      const base64string = "Basic " + buff.toString("base64");
      const reqOptions = {
        headers: { Authorization: base64string, "Content-Type": "text/xml", Depth: "1" },
        method: "get",
        url: caldavUrl + '?export&accept=jcal&start=' + startTimestamp + '&end=' + endTimestamp + '&expand=1'
      };
```

 When I send the request:

http://xxx.xxx.xxx.xxx:xxxx/dav.php/calendars/136worple/twickenhamevent?export&accept=jcal&start=1650294379&end=1681826779&expand=1&username=xxxxxxx&pass=xxxxxxx

I get:

```
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
    <s:sabredav-version>4.3.1</s:sabredav-version>
    <s:exception>Sabre\VObject\ParseException</s:exception>
    <s:message>This parser only supports VCARD and VCALENDAR files</s:message>
</d:error>
```

However, if I don't use expand, start and end like so:

http://xxx.xxx.xxx.xxx:xxxx/dav.php/calendars/136worple/twickenhamevent?export&accept=jcal&username=xxxxxxx&pass=xxxxxxx

I get JSON back. An example chunk:

`[
    "vcalendar",
    [
        [
            "calscale",
            {},
            "text",
            "GREGORIAN"
        ],
        [
            "version",
            {},
            "text",
            "2.0"
        ],
        [
            "prodid",
            {},
            "text",
            "-//SabreDAV//SabreDAV 4.3.1//EN"
        ],
        [
            "x-wr-calname",
            {},
            "unknown",
            "twickenhamevent"
        ],
        [
            "x-apple-calendar-color",
            {},
            "unknown",
            ""
        ]
    ],
    [
        [
            "vtimezone",
            [
                [
                    "tzid",
                    {},
                    "text",
                    "Europe/London"
                ],
                [
                    "x-lic-location",
                    {},
                    "unknown",
                    "Europe/London"
                ]
            ],
            [
                [
                    "daylight",
                    [
                        [
                            "tzname",
                            {},
                            "text",
                            "BST"
                        ],`

**Comment by ronschaeffer** [comment]
@kdw2060 No luck adding the `"Accept": "application/json"` header.

I've tried a few different searches on the issues at [sabre-io/dav](https://github.com/sabre-io/dav/issues) but haven't found anything helpful. There's[ this similar issue](https://github.com/sabre-io/dav/issues/1311) but there are no comments.

@cellerich Somewhere in the [sabre/dav docs](https://sabre.io/dav/) that I can't now find there is mention of various clients requiring variants of the URL beyond just omitting the trailing /. I'm probably grasping at straws, but could you share your calendar URL structure?

**Comment by zsamiatt** [comment]
Hello, I try using caldav (baikal server). No error in log but the sensor data emply. 

added 60 packages, and audited 61 packages in 29s
23 packages are looking for funding
  run `npm fund` for details
found 0 vulnerabilities
Calendar(s) queried at: Wed Aug 10 2022 16:30:00 GMT+0200 (Central European Summer Time)
CRON-Task: Events posted to sensor(s) at: Wed Aug 10 2022 16:30:00 GMT+0200 (Central European Summer Time)
Calendar(s) queried at: Wed Aug 10 2022 17:00:00 GMT+0200 (Central European Summer Time)
CRON-Task: Events posted to sensor(s) at: Wed Aug 10 2022 17:00:00 GMT+0200 (Central European Summer Time)
Calendar(s) queried at: Wed Aug 10 2022 17:30:00 GMT+0200 (Central European Summer Time)
CRON-Task: Events posted to sensor(s) at: Wed Aug 10 2022 17:30:00 GMT+0200 (Central European Summer Time)

**Comment by kdw2060** [comment]
Closing this issue. I am looking for someone who uses Baikal and who is willing to debug + maintain the baikal-specific code.

**Comment by alfureu** [comment]
Hm, I would like to reopen this, as I am running into issues with baikal. Other clients work fine, but HaOS somehow cannot make a successful pull neither with CalDAV nor with this addon.

The logs:
```
Add-on: Hass Calendar Addon
 Addon to consume caldav and google calendars and make calendar events available as sensordata in Home Assistant.
-----------------------------------------------------------
 Add-on version: 0.501
 You are running the latest version of this add-on.
 System: Home Assistant OS 16.2  (amd64 / qemux86-64)
 Home Assistant Core: 2025.9.4
 Home Assistant Supervisor: 2025.09.0
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums or the Discord chat.
-----------------------------------------------------------
s6-rc: info: service base-addon-banner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service base-addon-log-level: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service base-addon-log-level successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
/app/node_modules/undici/lib/web/webidl/index.js:531
webidl.is.File = webidl.util.MakeTypeAssertion(File)
                                               ^
ReferenceError: File is not defined
    at Object.<anonymous> (/app/node_modules/undici/lib/web/webidl/index.js:531:48)
    at Module._compile (node:internal/modules/cjs/loader:1364:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1422:10)
    at Module.load (node:internal/modules/cjs/loader:1203:32)
    at Module._load (node:internal/modules/cjs/loader:1019:12)
    at Module.require (node:internal/modules/cjs/loader:1231:19)
    at require (node:internal/modules/helpers:177:18)
    at Object.<anonymous> (/app/node_modules/undici/lib/web/fetch/util.js:12:20)
    at Module._compile (node:internal/modules/cjs/loader:1364:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1422:10)
Node.js v18.20.3
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service base-addon-log-level: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service base-addon-log-level successfully stopped
s6-rc: info: service base-addon-banner: stopping
s6-rc: info: service base-addon-banner successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```
